rfcs/001 - EC2 schema.md

```diff
@@ -0,0 +1,113 @@
+- Placement groups
+- Network interfaces
+
+Almost all the features listed will need a direct integration with the `instance` table, needing to add at least eight more columns to the model. Those columns are only for integration, but the `instance` model still needs more properties to be defined to be complete. Some of those missing properties are: instance auto-recovery, shutdown behaviour, stop - hibernate behaviour, termination protection, stop protection, detailed cloudWatch monitoring, elastic GPU, credit specification, tenancy, RAM disk ID, kernel ID or Metadata accessible (I might be missing others).
+
+The current instance module has ten columns. Adding the ones representing integration with other features will almost double its size.
+
+## Proposal
+
+The `instance` table is the core of the `aws_ec2` module and is the one that can grow more in complexity. The proposal is to keep adding the necessary columns to the `instance` table. It is better to have a unique table to look for instances than increase the complexity of the model, creating other tables and adding to the users the need to understand our model first. Eventually, we can add an `aws_ec2_simplified` module that can expose the basics to create an instance and leave the `aws_ec2` for advanced users.
+
+We can divide the feature list as follow:
+
+The direct implication in the `instance` table:
+
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Capacity reservations
+- Auto Scaling Groups
+- Elastic IP integration
+- Placement groups
+- Network interfaces
+
+No implication in the `instance` table:
+
+- Instance type - Availability zone relationship
+- AMIs management
+- EBS volume snapshots and life cycle
+
+The `Saving plans`, `Reserved instances` or `Scheduled instances` are not included in the list above since they seem to be budgeting solutions that need interaction with the user rather than infrastructure-related modules' features that could have any implication in the model.
+
+Let's start with the features without impact on the `instance` table. These features will be part of the `aws_ec2` module with their mappers. The Instance type table by availability zones will be a table with a composite primary key. It technically has a minimum impact in the `instance` table that will be a special FK to the composite primary key, but no UX model impact.
+
+The features with direct implications in the `instance` table can be divided into two groups: one group affects only the `aws_ec2` module, the other impacts more modules.
+
+Let's start with the first group, Placement groups and Network interfaces. These features will be part of their respective modules (probably `aws_vpc`?). Then, they will need to be integrated with the `instance` table in the `aws_ec2` module with an FK nullable column.
+
+The rest of the features will need to be implemented in the `aws_ec2` module and integrated with the `instance` table in the `aws_ec2` module with an FK nullable column.
```

Comment:
so currently `registered_instances` has the FK outside the `instance` table. would it be better to keep the UX of things consistent or do we have a strong reason to keep that different to these tables? do we anticipate any other tables like `registered_instances`?

Reply:
Well I think that specific case could not be modeled otherwise, is a `many-to-many` relationship, where an instance can be registered with multiple target groups and target groups can have multiple ec2 instances registered.

The network interfaces would be a similar case to ebs volumes, where an instance can have several ENIs but one ENI is only assigned to an instance, so the FK to the instance will be in the network interface table.